nixos/modules/services/networking/rosenpass.nix

```
@@ -0,0 +1,220 @@
+            such a scenario is unusual. Please double-check your configuration.
+          '';
+        in
+        (optional (relevantExtractions != [ ] && missingKeys != [ ]) ''
+          You have configured Rosenpass peers with the WireGuard public keys:
+          ${itemize missingKeys}
+          But there is no corresponding active Wireguard peer configuration in any of:
+          ${itemize (descriptions relevantExtractions)}
+          ${unusual}
+        '')
+        ++
+        (optional (relevantExtractions == [ ]) ''
+          You have configured Rosenpass, but you have not configured Wireguard via any of:
+          ${itemize (descriptions extractions)}
+          ${unusual}
+        '');
+
+      environment.systemPackages = [ cfg.package pkgs.wireguard-tools ];
+
+      systemd.services.rosenpass =
+        let
+          filterNonNull = filterAttrsRecursive (_: v: v != null);
+          config = settingsFormat.generate "config.toml" (
+            filterNonNull (settings
+              //
+              (
+                let
+                  credentialPath = id: "$CREDENTIALS_DIRECTORY/${id}";
+                  # NOTE: We would like to remove all `null` values inside `settings`
+                  # recursively, since `settingsFormat.generate` cannot handle `null`.
+                  # This would require to traverse both attribute sets and lists recursively.
+                  # `filterAttrsRecursive` only recurses into attribute sets, but not
+                  # into values that might contain other attribute sets (such as lists,
+                  # e.g. `settings.peers`). Here, we just specialize on `settings.peers`,
+                  # and this may break unexpectedly whenever a `null` value is contained
+                  # in a list in `settings`, other than `settings.peers`.
+                  peersWithoutNulls = map filterNonNull settings.peers;
+                in
+                {
+                  secret_key = credentialPath "pqsk";
+                  public_key = credentialPath "pqpk";
+                  peers = peersWithoutNulls;
+                }
+              )
+            )
+          );
+        in
+        rec {
+          wantedBy = [ "multi-user.target" ];
+          after = [ "network-online.target" ];
+          path = [ cfg.package pkgs.envsubst pkgs.wireguard-tools ];
+
+          serviceConfig = {
+            User = "rosenpass";
+            Group = "rosenpass";
+            RuntimeDirectory = "rosenpass";
+            DynamicUser = true;
+            AmbientCapabilities = [ "CAP_NET_ADMIN" ];
+            LoadCredential = [
+              "pqsk:${settings.secret_key}"
+              "pqpk:${settings.public_key}"
+            ];
+          };
+
+          # See <https://www.freedesktop.org/software/systemd/man/systemd.unit.html#Specifiers>
+          environment.CONFIG = "%t/${serviceConfig.RuntimeDirectory}/config.toml";
+
+          preStart = "envsubst -i ${config} -o \"$CONFIG\"";
```

```suggestion
          preStart = "${lib.getExe envsubst} -i ${config} -o \"$CONFIG\"";
```
easy way to keep PATH cleaner